src/server/zset_family.cc

```diff
@@ -948,6 +957,23 @@ OpResult<AddResult> OpAdd(const OpArgs& op_args, const ZParams& zparams, string_
   AddResult aresult;
   detail::RobjWrapper* robj_wrapper = res_it.value()->second.GetRobjWrapper();
   bool is_list_pack = robj_wrapper->encoding() == OBJ_ENCODING_LISTPACK;
+
+  // opportunistically reserve space if multiple entries are about to be added.
+  if ((zparams.flags & ZADD_IN_XX) == 0 && members.size() > 2) {
+    if (is_list_pack) {
+      uint8_t* zl = (uint8_t*)robj_wrapper->inner_obj();
+      size_t malloc_reserved = zmalloc_size(zl);
+      size_t min_sz = EstimateListpackMinBytes(members);
+      if (min_sz > malloc_reserved) {
+        zl = (uint8_t*)zrealloc(zl, min_sz);
```

Comment:
If this fails we will end up setting the `inner_obj` to `nullptr`

Reply:
Not sure I understand under what conditions zrealloc can fail . In any case, listpack grows by calling zrealloc (see listpack.c), so we are as good as all other listpack functions.

Reply:
Your previous comment I think covers it, so it should not fail even if it fails to reallocate (it just returns a new block). It's fine for now and if we end up having an issue we are just gonna fix it (which based on what you say there shouldn't be any)